articles/121_realtime_proposal.md

```diff
@@ -0,0 +1,215 @@
+Here are some features critical to deterministic performance that are often exposed to the user in an RTOS, but not in a conventional OS (see Wikipedia, [Real-time operating system](https://en.wikipedia.org/wiki/Real-time_operating_system)).
+For more information about why these features matter, refer to the [introductory article on real-time systems](http://design.ros2.org/articles/realtime_background.html).
+
+- Task scheduling
+
+  - Expose thread priority and deterministic schedulers that incorporate priority
+  - Allow full thread preemption
+
+- Concurrency and synchronization primitives
+
+  - Floored semaphores (allow a higher priority thread to take control of locked semaphore)
+
+- Interrupt handling
+
+  - Disable interrupts
+  - Mask interrupts (temporarily suspend interrupts in a critical section)
+
+- Memory allocation
+
+Some real-time systems have a non real-time component, e.g. for graphical display or other user interaction.
+This would require an operating system that allows non real-time threads to be scheduled alongside real-time threads.
+Because of the extra nondeterminism introduced by this requirement, this kind of system would be better suited to a soft real-time constraint.
+
+Another important constraint is which processor architectures are supported by which operating systems, and which RTOS is an appropriate choice due to performance constraints of the architecture.
+
+## Goal for ROS 2
+
+In order to support real-time systems in ROS 2, we propose an iterative approach for identifying limitations in the API that force the code path into performance bottlenecks.
+
+1. Select two use cases with real-time requirements that represent two extremes of the space described by diagram (#), for example:
+
+  - An STM32 embedded board running no operating system or a hard real-time OS with 1 kHZ update frequency with 3% allowable jitter.
+  - An x86 computer running Xenomai or RT_PREEMPT that receives the sensor data over Ethernet or USB and processes the output in soft or firm real-time, but also runs separate non-RT processes for user interaction.
+
+2. Implement the system with ROS 2.
+3. Profile the performance of the system.
+4. Identify solutions to improve the performance of the system, e.g. by exposing more concurrency primitives through the appropriate abstraction.
+5. Adapt the ROS 2 API or underlying libraries to implement these solutions and to improve the experience of writing real-time code.
+6. Refactor the system for the new API if necessary and repeat.
+
+## Software architecture challenges and solutions
+
+Wherever ROS 2 libraries expose sources of nondeterminism, there should be a path that exposes an alternative.
+This section evaluates the current state of ROS 2 and rclcpp, predicts challenges to fulfilling real-time requirements, and presents possible solutions.
+
+### Memory Allocation
+
+As of the writing of this document (1/6/16), selecting custom allocators in rclcpp is almost fully implemented.
```

Comment:
An introductory sentence about what custom allocators are and why they are relevant here may be nice.